### PR TITLE
Feature/tx fees 2.0

### PR DESCRIPTION
**Pull request by Jedi2002**
## **Transaction Fees 2.0**
 (**not** "first swap on us")

**HOW IT WORKS:**
- First we create a signed doc and sign a transaction on the user’s behalf (e.g. Alice). Sign transaction forces the regular Keplr popup to appear, and Alice is prompted to select the transaction fee and gas amount as usual. The resulting transaction fee is exactly what Alice is going to pay when the transaction takes place (e.g. is actually broadcasted).
- Right after that (behind the scenes), the broadcast transaction takes place. It uses a previously signed transaction as an input. At this point Alice still does not have ROWAN to pay the transaction fee and this is when the BE takes it over and makes sure that Alice has enough ROWAN before the transaction is delivered. 
- BE handles various edge cases like if Alice has 0 ROWAN or 50% of the required amount, or has more than enough ROWAN to pay transaction fees. 

- Alice will pay exactly what was selected in the Keplr popup in the first step when signing the transaction and BE adds some ROWAN to the existing amount (if any) so the sum is exactly what is required for tx. We do not cover tx fees for free. Instead, BE subtracts the corresponding amount from the input token, hence Alice receives a bit less of the output token.

### CODE REVIEW

**Review by iverc — APPROVED** [review]

**Review by khdegraaf — CHANGES_REQUESTED** [review]
Tests are still failing, but integration tests aren't always running in each push, in particular if you are only editing UI code.

**Review by khdegraaf — APPROVED** [review]
Integration tests now pass, so no issues there anymore...

**Review by iverc — DISMISSED** [review]

File: app/upgrade-poolChange.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/Sifchain/sifnode/x/clp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/upgrade"
+	"io/ioutil"
+)
+
+func GetPoolChangeFunc(app *SifchainApp) func(ctx sdk.Context, plan upgrade.Plan) {
```
Comment (Jedi2002) [inline app/upgrade-poolChange.go:12]: Just changed the location ,the code in this file is old ,and already reviewd

File: ui/core/src/entities/formulae.calculateDoubleSwapResult.test.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import { calculateExternalExternalSwapResult } from "./formulae";
```
Comment (Jedi2002) [inline ui/core/src/entities/formulae.calculateDoubleSwapResult.test.ts:1]: UI code ,not relevent to the backend implementation

File: ui/core/src/entities/formulae.calculateReverseSwapResult.test.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { Fraction, TEN } from "./fraction/Fraction";
```
Comment (Jedi2002) [inline ui/core/src/entities/formulae.calculateReverseSwapResult.test.ts:1]: UI code ,not relevent to the backend implementation

File: app/ante.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+import (
+	"github.com/Sifchain/sifnode/x/clp"
+	"github.com/Sifchain/sifnode/x/faucet"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth"
+	"github.com/cosmos/cosmos-sdk/x/auth/ante"
+	"github.com/cosmos/cosmos-sdk/x/auth/types"
+)
+
```
Comment (alexanderbez) [inline app/ante.go:11]: nit: I would add godocs to these ante-handler functions just so the reader knows the difference between each. They don't need to be too verbose.

File: app/upgrade-poolChange.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+			if !app.clpKeeper.ValidatePool(pool) {
+				hasError = true
+				ctx.Logger().Error(fmt.Sprintf("Invalid | Pool %s ", pool.String()))
+				break
+			}
+			poolList = append(poolList, pool)
+		}
+		// If we have error dont set state
+		if hasError {
+			ctx.Logger().Error("Failed to execute upgrade plan for pool re-balance")
```
Comment (alexanderbez) [inline app/upgrade-poolChange.go:58]: Just return here and then you can avoid the `!hasError` check below.

File: app/upgrade-poolChange.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+		poolList := clp.Pools{}
+		hasError := false
+		for _, pool := range allPools {
+			lpList := app.clpKeeper.GetLiquidityProvidersForAsset(ctx, pool.ExternalAsset)
+			temp := sdk.ZeroUint()
+			tempExternal := sdk.ZeroUint()
+			tempNative := sdk.ZeroUint()
+			for _, lp := range lpList {
+				withdrawNativeAssetAmount, withdrawExternalAssetAmount, _, _ := clp.CalculateWithdrawal(pool.PoolUnits, pool.NativeAssetBalance.String(),
+					pool.ExternalAssetBalance.String(), lp.LiquidityProviderUnits.String(), sdk.NewUint(clp.MaxWbasis).String(), sdk.NewInt(0))
```
Comment (alexanderbez) [inline app/upgrade-poolChange.go:29]: nit: a bit difficult to read
```suggestion
				withdrawNativeAssetAmount, withdrawExternalAssetAmount, _, _ := clp.CalculateWithdrawal(
				  pool.PoolUnits, 
				  pool.NativeAssetBalance.String(),
				  pool.ExternalAssetBalance.String(), 
				  lp.LiquidityProviderUnits.String(), 
				  sdk.NewUint(clp.MaxWbasis).String(), 
				  sdk.NewInt(0),
				)
```

File: app/upgrade-poolChange.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+				newLpUnits, lpUnits, err := clp.CalculatePoolUnits(pool.ExternalAsset.Symbol, temp, tempNative, tempExternal,
+					withdrawNativeAssetAmount, withdrawExternalAssetAmount)
```
Comment (alexanderbez) [inline app/upgrade-poolChange.go:31]: ditto
```suggestion
				newLpUnits, lpUnits, err := clp.CalculatePoolUnits(
				  pool.ExternalAsset.Symbol,
				   temp,
				    tempNative,
				     tempExternal,
				  withdrawNativeAssetAmount,
				  withdrawExternalAssetAmount,
				  )
```

File: app/upgrade-poolChange.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+	err = ioutil.WriteFile(fmt.Sprintf("%v-state.json", name), appStateJSON, 0600)
+	if err != nil {
+		ctx.Logger().Error(fmt.Sprintf("failed to write state to file: %s", err.Error()))
+	}
+	err = ioutil.WriteFile(fmt.Sprintf("%v-validator.json", name), valList, 0600)
+	if err != nil {
+		ctx.Logger().Error(fmt.Sprintf("failed to write Validator List to file: %s", err.Error()))
+	}
```
Comment (alexanderbez) [inline app/upgrade-poolChange.go:95]: Why do you not want to preemptively return an any non-nil error case?

File: x/clp/alias.go [hunk 9]
```diff
@@ -15,6 +15,8 @@ const (
 	MaxWbasis         = types.MaxWbasis
 	PoolThrehold      = types.PoolThrehold
 	PoolUnitsMinValue = types.PoolUnitsMinValue
+	SwapType          = types.SwapType
+	TxFeeMultiplier   = types.TxFeeMultiplier
 )
```
Comment (alexanderbez) [inline x/clp/alias.go:21]: Just a friendly reminder, that the SDK has moved away from the `alias.go` file convention -- it really didn't add any benefits. That being said, feel free to keep it if it makes sense for your team.

File: x/faucet/ante.go [hunk 10]
```diff
@@ -0,0 +1,23 @@
+package faucet
+
+import (
+	"github.com/Sifchain/sifnode/x/clp"
+	types2 "github.com/Sifchain/sifnode/x/faucet/types"
```
Comment (alexanderbez) [inline x/faucet/ante.go:5]:
```suggestion
	ftypes "github.com/Sifchain/sifnode/x/faucet/types"
```

File: x/faucet/ante.go [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+type RemoveFaucetFeeDecorator struct{}
+
+func NewRemoveFacuetFeeDecorator() RemoveFaucetFeeDecorator {
+	return RemoveFaucetFeeDecorator{}
+}
+
+var _ types.AnteDecorator = RemoveFaucetFeeDecorator{}
+
+func (r RemoveFaucetFeeDecorator) AnteHandle(ctx types.Context, tx types.Tx, simulate bool, next types.AnteHandler) (newCtx types.Context, err error) {
```
Comment (alexanderbez) [inline x/faucet/ante.go:17]: I would add a small godoc here illustrating what is happening. e.g. `// On a testnet only where the chain-id is not equal to "sifchain" ...`

File: x/faucet/handler.go [hunk 12]
```diff
@@ -67,8 +67,8 @@ func handleMsgRequestCoins(ctx sdk.Context, keeper Keeper, msg types.MsgRequestC
 // Handle the add coins message and send coins from the signers account to the module account.
 func handleMsgAddCoins(ctx sdk.Context, keeper Keeper, msg types.MsgAddCoins) (*sdk.Result, error) {
 	if ctx.ChainID() != "sifchain" {
-		bank := keeper.GetBankKeeper()
-		err := bank.SendCoins(ctx, msg.Signer, types.GetFaucetModuleAddress(), msg.Coins)
+		supply := keeper.GetSupplyKeeper()
```
Comment (alexanderbez) [inline x/faucet/handler.go:70]: Why the change in name? The bank keeper is still used, no?

File: x/clp/ante.go [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+	msg := feeTx.GetMsgs()[0]
+	switch msg := msg.(type) {
+	case clpTypes.MsgSwap:
+		payer := feeTx.FeePayer()
+		// Fee Payer is always the msg signer for swap transactions
+		if !payer.Equals(msg.Signer) {
+			return types.Context{}, errors.New("Fee Payer and MSG Signer are not the same ")
+		}
+		// User is sending rowan , that means they already have rowan balance
+		if msg.SentAsset.Equals(GetSettlementAsset()) {
```
Comment (alexanderbez) [inline x/clp/ante.go:39]: What's the difference between `clpTypes.GetSettlementAsset()` and `GetSettlementAsset()`?

File: x/clp/ante.go [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+			return next(ctx, tx, simulate)
+		}
+		// The amount of Fee user agreed to pay
+		feeInRowan := feeTx.GetFee()
+		requiredRowan := feeInRowan.AmountOf(clpTypes.GetSettlementAsset().Symbol)
```
Comment (alexanderbez) [inline x/clp/ante.go:44]: You call `clpTypes.GetSettlementAsset().Symbol` twice. May I suggest storing it in a variable?

File: app/upgrade-poolChange.go (hunk 5 above)
Comment (Jedi2002) [reply app/upgrade-poolChange.go:70]: Yeah , this can be added in , but this is an older upgrade which already went through .

File: app/upgrade-poolChange.go (hunk 8 above)
Comment (Jedi2002) [reply app/upgrade-poolChange.go:107]: I think the idea was, to write to a whatever we can , becuase I see ExportAppState does not return an error . 
So its like write to a file if you can , if you cant just continue . 
@iverc , is that correct ?

File: x/faucet/handler.go (hunk 12 above)
Comment (Jedi2002) [reply x/faucet/handler.go:70]: The refactor is just to mainitain consistence ,the request and add coins now use similar functions .

**Review by iverc — APPROVED** [review]

**Comment by intl-man** [comment]
@Jedi2002 The UI has since been decoupled. If this is still relevant, then a new PR should be raised. Let me know, thanks!
